File: blaise-graph-theory/src/main/java/com/googlecode/blaisemath/graph/metrics/AbstractGraphNodeMetric.java
package com.googlecode.blaisemath.graph.metrics;

/*
 * #%L
 * blaise-graphtheory3
 * --
 * Copyright (C) 2009 - 2019 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.google.common.graph.Graph;
import com.googlecode.blaisemath.graph.GraphNodeMetric;

import java.util.Map;

import static java.util.stream.Collectors.toMap;

/**
 * Partial implementation of a node metric that gives a name to the metric and provides a default implementation for
 * mapping all values.
 * 
 * @param <T> metric result type
 *
 * @author Elisha Peterson
 */
public abstract class AbstractGraphNodeMetric<T> implements GraphNodeMetric<T> {
   
    protected final String name;

    public AbstractGraphNodeMetric(String name) {
        this.name = name;
    }
    
    @Override
    public String toString() {
        return name;
    }

    @Override
    public <N> Map<N, T> apply(Graph<N> graph) {
        return graph.nodes().stream().collect(toMap(n -> n, n -> apply(graph, n)));
    }
    
}

File: blaise-graph-theory/src/main/java/com/googlecode/blaisemath/graph/metrics/AbstractGraphNodeMetric.kt
package com.googlecode.blaisemath.graph.metrics

import com.google.common.graph.Graph
import com.googlecode.blaisemath.graph.ContractedGraphTest
import com.googlecode.blaisemath.graph.GraphNodeMetric
import com.googlecode.blaisemath.graph.GraphUtilsTest
import com.googlecode.blaisemath.graph.SubgraphTest
import com.googlecode.blaisemath.graph.generate.GraphGrowthRule
import com.googlecode.blaisemath.graph.generate.GraphSeedRule
import com.googlecode.blaisemath.graph.generate.HopGrowthRule
import com.googlecode.blaisemath.graph.layout.SpringLayoutPerformanceTest
import com.googlecode.blaisemath.graph.metrics.AdditiveSubsetMetricTest
import com.googlecode.blaisemath.graph.metrics.BetweenCentralityTest
import com.googlecode.blaisemath.graph.metrics.ClosenessCentralityTest
import com.googlecode.blaisemath.graph.metrics.CooperationMetric
import com.googlecode.blaisemath.graph.metrics.EigenCentralityTest
import com.googlecode.blaisemath.graph.metrics.GraphCentralityTest
import com.googlecode.blaisemath.graph.metrics.SubsetMetricsTest
import com.googlecode.blaisemath.test.AssertUtils
import org.junit.BeforeClass
import java.util.stream.Collectors

/*
* #%L
* blaise-graphtheory3
* --
* Copyright (C) 2009 - 2021 Elisha Peterson
* --
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* #L%
*/ /**
 * Partial implementation of a node metric that gives a name to the metric and provides a default implementation for
 * mapping all values.
 *
 * @param <T> metric result type
 *
 * @author Elisha Peterson
</T> */
abstract class AbstractGraphNodeMetric<T>(protected val name: String?) : GraphNodeMetric<T?> {
    override fun toString(): String {
        return name
    }

    override fun <N> apply(graph: Graph<N?>?): MutableMap<N?, T?>? {
        return graph.nodes().stream().collect(Collectors.toMap({ n: N? -> n }) { n: N? -> apply(graph, n) })
    }
}